Iterate by chunks over specified axis

vbarrielle opened the pull request:
This relies on iterating using `OuterIterCore` while monitoring the location of the current index. When the iterator detects the last chunk, a different shape is used.

I'm not sure this is the proper design yet, it works for forward iteration but probably not for reversed iteration.

vbarrielle commented:
Reversed iteration works with the last changes. I'll proof read that again later but it should be ok.

bluss commented on tests/iterators.rs:
```diff
@@ -295,6 +295,55 @@ fn axis_iter_mut() {
+    let a = a.reshape((2, 6, 2));
+
+    let mut it = a.axis_chunks_iter(1, 2);
+    assert_eq!(it.next().unwrap(),
```
Using `assert_equal` would make these asserts nicer. See tests above.

bluss commented on src/iterators.rs:
```diff
@@ -641,3 +641,157 @@ pub fn new_axis_iter_mut<A, D>(v: ArrayViewMut<A, D>,
+fn chunk_iter_parts<A, D: Dimension>(v: ArrayView<A, D>,
+                                     axis: usize,
+                                     size: usize
+                                    ) -> (OuterIterCore<A, D>, *mut A, D)
```
I can't really handle these hanging `)` and returns. Sure, the style in the crate is not super consistent but it does not use these.

vbarrielle replied:
Ok, I'll try not to use that style. Btw maybe there should be a rustfmt file for the crate at some point?

bluss replied:
I've never been happy with rustfmt, maybe I should try with the newest version!

bluss commented:
Reverse iteration doesn't need to be a goal, only when it's cost free to do so.

bluss commented on src/iterators.rs:
```diff
@@ -641,3 +641,157 @@ pub fn new_axis_iter_mut<A, D>(v: ArrayViewMut<A, D>,
+///
+/// For example, in a 2 × 8 × 3 array, if the axis of iteration
+/// is 1 and the chunk size is 2, the yielded elements
+/// are 2 × 2 × 3 subviews (and there are 4 in total).
```
we can call them views instead of "subviews" since they have the same dimensionality as the original array.

bluss replied:
Good example text otherwise! :smile:

bluss commented:
Oh, my review question that I haven't managed to answer myself is, does the raw pointer of the iteration stay in bounds? We cannot have it point outside the array's allocation, even when we just use the last value to stop on.  I think it should be ok, by the construction of OuterIterCore.

vbarrielle commented:
Yes the raw pointer should always be in bounds because the stride and length given to OuterIterCore ultimately point to the last block and not any further. Thus the only additional risk is if the last shape is not correct.

bluss commented on src/lib.rs:
```diff
@@ -1432,6 +1434,47 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
+    /// assert_eq!(2, iter.next().unwrap().shape()[1]);
+    /// assert_eq!(2, iter.next().unwrap().shape()[1]);
+    /// assert_eq!(2, iter.next().unwrap().shape()[1]);
+    /// assert_eq!(1, iter.next().unwrap().shape()[1]);
```
This example is unfortunately too cryptic. We can show just the first view in the iterator instead. Or some pretty ascii pictures... (I can add that if nothing else comes up)

vbarrielle commented:
The latest changes should address your remarks. Though for the style change I'm not sure whether the style I've used suits you.

bluss commented:
Beautiful!

I'm back at my review question. I think `last_ptr` can point outside of the array, which we can't do unfortunately.

Consider the case where `size` equals the length of the axis. last_ptr will point outside of the array.

For a regular slice we can point "one past the end" but no pointing further outside is legal (when using `ptr.offset`)
```
[a, b, c, d] x
 ^  ^  ^  ^  ^    valid pointer targets in and next to a slice
```

With a non-unitary stride, we easily land way past the end (or before it, if it's negative). We may not point to one row past the end.

bluss commented:
To clarify, it is UB to use `ptr.offset` to point at anything but the “valid pointer targets”, even if we don't deref the pointer.

bluss commented:
Doing that safely is a digression for another day (and when this is stable) http://doc.rust-lang.org/nightly/std/intrinsics/fn.arith_offset.html

vbarrielle commented:
> To clarify, it is UB to use ptr.offset to point at anything but the “valid pointer targets”, even if we don't deref the pointer.

I see, I didn't know it was UB, I thought only derefing was UB. Looks like I need to add some special case for when `size == self.dim[axis]`

bluss commented:
Additionally maybe clamp size to the length of the dimension if needed. Silently accepting a larger size is consistent with the uneven last view anyway.

vbarrielle commented:
Right that looks like a good behaviour.

vbarrielle commented:
I've thought about that offseting a bit more. With `size` clamper to the dimension length, there is an invariant `last_index * iterator_stride <= dim_len * dim_stride`, and the inequality is an equality when `size == dim_len`. So at worst the offset will point to right after the last allocated element if I'm correct.

This is however subject to the assumption that, for a dimension of length 1, the stride does not go out of bounds. It should be enforced by the strides checks but I'm unsure about this special case.

However we're better sure than sorry, so I've implemented the special handling for `size == dim_len`.

bluss commented:
That assumption does not hold for views of length 1. Consider `[0, 1, 2, 3]` sliced like `1..; 2`; it has elements `[1, 3], len = 2, stride = 2` the "one past the end" in stride * dim_len = 4 terms is outside of the valid range.

We don't know where the last allocated element is (stride may be negative). I don't see any other way but making sure the pointer always points at something we know is a valid element.

bluss commented on src/iterators.rs:
```diff
@@ -641,3 +640,157 @@ pub fn new_axis_iter_mut<A, D>(v: ArrayViewMut<A, D>,
+
+    let last_ptr = if last_index < axis_len {
+        unsafe {
+            v.ptr.offset(stride * last_index as isize)
```
I don't want to tire you out, but if `size == axis_len` here, we still compute an offset outside of the array, because `stride = axis_stride * size`.

I guess the condition `size < axis_len` would work well?

vbarrielle replied:
Don't worry you won't tire me out, I want this to be safe too :)
It's probable `size < axis_len` works, I'll test that against the examples you provided later.

vbarrielle replied:
Actually this won't work either, I think there's a possibility to index out of bounds if `rem == 0`. That's because any time the size divides the dimension length, `last_index` is not a "valid" index (ie `last_index * stride >= axis_len`).

So it seems the proper test is `rem != 0`.

bluss commented:
Here's ~~a~~two demos, where last_ptr ends up pointing outside the range. Just to make it a bit more concrete.

```rust
    let a = Array::<f32, _>::zeros((8, 1));
    let sl = a.as_slice().unwrap();
    println!("pointers: {:p}, end={:p}", sl.as_ptr(),
        unsafe { sl.as_ptr().offset(sl.len() as isize) });
    let a = a.slice(s![..;-1,..]);
    let it = a.axis_chunks_iter(0, 8);


    let a = Array::<f32, _>::zeros((8, 2));
    let sl = a.as_slice().unwrap();
    println!("pointers: {:p}, end={:p}", sl.as_ptr(),
        unsafe { sl.as_ptr().offset(sl.len() as isize) });
    let a = a.slice(s![1..;2,..]);
    let it = a.axis_chunks_iter(0, 8);
```

vbarrielle commented:
I've added a better check to avoid offseting out of bounds. Checking `rem != 0` is the proper check imho. I've added more diverse test cases and can no longer witness an offseting going out of bounds.

bluss commented:
Yep I agree. Now it's good :smile:

bluss commented:
:fish_cake:

vbarrielle commented:
Cool, thanks for your help with this, I'm not yet used to the requirements of unsafe rust, so it was nice having you pointing out the flaws.

bluss commented:
Yep, no problem, it's great that we can work on it like this. Rust didn't give us the largest challenges in the world this time..
